datafusion/expr-common/src/columnar_value.rs

```diff
@@ -89,7 +91,7 @@ pub enum ColumnarValue {
     /// Array of values
     Array(ArrayRef),
     /// A single value
-    Scalar(ScalarValue),
+    Scalar(Scalar),
```

Comment:
when i read @jayzhan211 's https://github.com/apache/datafusion/pull/12488#issuecomment-2359844558 o thought that ScalarValue is going to be left in ColumnarValue (least overhead, as relating to @alamb 's https://github.com/apache/datafusion/pull/12488#pullrequestreview-2312833258).
But here is't being replaced.

What's the next step here? what's the end state?

Reply:
The next step is to slowly remove variants of `ScalarValue` while still accounting for them via `data_type` (which in my opinion is the least intrusive change to support this effort). After we are satisfied with the variants that remain we reconsider the logic in expressions and operators in order to support what @jayzhan211 in proposing:

> I think we could have LogicalType without any Arrow's DataType contained in it in the future

by sourcing the `data_type` directly from the execution schema.

Reply:
- Is ScalarValue supposed to be in the logical types layer, or physical?
- is Scalar supposed to be in the logical types layer, or physical?
- is ColumnarValue purely physical layer?

> remove variants of `ScalarValue` while still accounting for them via `data_type`

what do we need this for?
(maybe that's obvious for someone with clarity on the preceding questions)

Reply:
I think it is hard to tell the role given we are in the intermediate state

The end state in my mind is that,
`ColumnarValue` stores `ArrayRef` for multiple rows and `ScalarValue` for single rows case.
We will have `LogicalType` which could be resolved from `DataType` given the mapping we determined, which should be customizable as well for user defined type. Similar to the role `Scalar` in this PR, which record the relationship between `ScalarValue` to `DataType`.

Reply:
> remove variants of ScalarValue while still accounting for them via data_type


Our first step is to minimize the variant of types, so we don't need `ScalarValue::Utf8`, `ScalarValue::LargeUtf8` any more but `Scalar` which has `ScalarValue::String` + `DataType::Utf8` or `ScalarValue::String` + `DataType::LargeUtf8`

Reply:
> The end state in my mind is that,
> `ColumnarValue` stores `ArrayRef` for multiple rows and `ScalarValue` for single rows case. `ScalarValue` has either native type like `i64` or `arrow::Scalar`.

Ok, so `ScalarValue` is going to be physical and also going to be simplified.

what will be used in `Expr::Literal`?
i am very interested in this part


> `ScalarValue::String` is then something close to `logical type`

but being used in very physical context, right?

Reply:
For value, there are `ScalarValue` and `ArrayRef`, for type there are `LogicalType` and `DataType`.

`Scalar` will be used in `Expr::Literal`, it contains `ScalarValue::String` + `DataType`. We can get the logical type from `DataType` or `ScalarValue` iteslf in the logical layer. And, decode the `String` with `DataType` in the physical layer.

Reply:
For `ScalarValue::List`, we might need something like `ScalarValue::List(Vec<ScalarValue>)` instead of `arrow::Scalar` 🤔

Reply:
`ScalarValue` should be used all over datafusion, not restricted to logical type or physical type. We just need to bring along with `DataType` so we understand how to decode the value.

Reply:
> , it contains `ScalarValue::String` + `DataType`

so a number is stored as "5"?
how do i compare two float scalar values? i need to parse them?

how do i represent a constant (aka "literal") of type `array<row<a: map<string, array<row<a: int, b: int>>>>>`?

i do believe constants in the plan are better expressed in their native representation, like we did in `ScalarValue` (or via `Any`, as long as we have type information needed to cast it back)

i guess this is a similar concern to one raised by @alamb here: https://github.com/apache/datafusion/pull/12536#discussion_r1774086062

Reply:
Number is not strictly stored as string. ScalarValue::I32(5) is numeric, ScalarValue::String("5") is numeric string. ScalarValue::Float(1.1) is numeric, you can compare them like rust native float type.

Reply:
I think there is nothing change for literal case, we still get `Literal(ScalarValue)`.

Reply:
> `ScalarValue` should be used all over datafusion, it is more a **value** and has no logical or physical concept. We just need to bring along with `DataType` so we understand how to decode the value in physical layer and understand what logical type it is to interact within the logical plan

And bring along `DataType` could be a transition state, I guess we could get the DataType from schema so we don't even require `DataType` in `ScalarValue`

Reply:
As long as we have `Schema` we can get the `DataType`, and we can get the LogicalType from it if each DataType has at most one LogicalType. `ScalarValue` is just a wrapper for the scalar case of `ArrayRef` or rust native type. We can carry `String` for `Utf8` / `Utf8View` / `LargeUtf8` and decode it if needed, and also carry `Scalar<ArrayRef>` for `List` / `LargeList` and so on.

Reply:
> we can get the LogicalType from it if each DataType has at most one LogicalType

We should not assume this to be the case.
I suppose https://github.com/apache/datafusion/issues/12644 may convey why, but even if we don't do Extension Types, this assumption will be very limiting.

Reply:
> > we can get the LogicalType from it if each DataType has at most one LogicalType
> 
> We should not assume this to be the case. I suppose #12644 may convey why, but even if we don't do Extension Types, this assumption will be very limiting.

I'm still not convinced why my assumption is limited. For extension types, we could customize/extend the **logical type** for each `DataType`. It doesn't make sense to have more than 1 logical type that decoded to the same physical type, they are *equivalent* thus should be the same logical type. Even if they have historical issue thus require more than one type for backward compatibility, they are able to do it if they built their type on top of Datafusion's Logical Type.